Support exposing the InspectorService over the Flutterservice extension protocol as well as the observatory protocol.

jacob314 opened the pull request:
We will probably remove most of the observatory protocol support once a
couple versions of the Flutter IntelliJ plugin have shipped that use the
Flutter service extension protocol. The only reason to continue supporting
the observatory protocol is it will allow using the inspector when paused
at a breakpoint.
Note tests are generally just versions of the existing tests for the observatory protocol support. I've erred on the side of duplicating test code vs making the test logic less obvious.

Matching flutter intellij CL.
https://github.com/flutter/flutter-intellij/commit/f64904446039970ac495f45fa40d06fb3734db2d

jacob314 committed:
Support exposing the InspectorService over the Flutter
service extension protocol as well as the observatory protocol.

We will probably remove most of the observatory protocol support once a
couple versions of the Flutter IntelliJ plugin have shipped that use the
Flutter service extension protocol. The only reason to continue supporting
the observatory protocol is it will allow using the inspector when paused
at a breakpoint.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -117,11 +117,14 @@ class _InspectorReferenceData {
 /// All methods returning String values return JSON.
 class WidgetInspectorService {
-  WidgetInspectorService._();
+  @visibleForTesting
```
this will need documentation.

BTW, @visibleForTesting doesn't seem to really do anything. So really this just makes it public.

jacob314 replied:
I've followed the suggestion to make this class only work as a mixin which seems reasonable given @visibleForTesting isn't doing anything really useful yet.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -146,10 +149,203 @@ class WidgetInspectorService {
+  /// property named `result` in the JSON. In case of failure, the failure is
+  /// reported to the remote caller and is dumped to the logs.
+  @visibleForTesting
+  void registerServiceExtension({
```
maybe this class should now just be a Binding?

Hixie replied:
I guess not, it's not its own library, it's tightly part of "widgets".

jacob314 replied:
Yes it is tightly coupled with widgets although making it part of binding is very tempting as there are a lot of similarities and some duplicated code.

Hixie commented on packages/flutter/lib/src/widgets/binding.dart:
```diff
@@ -295,6 +296,8 @@ abstract class WidgetsBinding extends BindingBase with SchedulerBinding, Gesture
         }
     );
+
+    WidgetInspectorService.instance.initServiceExtensions();
```
Seems like if we're doing this now, the previous one (debugWidgetInspector) might as well go there too?

jacob314 replied:
I've added 
ext.flutter.inspector.show which is identical to 
ext.flutter.debugWidgetInspector 
I can't remove debugWidgetInspector until users are updated to a version of the IntelliJ plugin that supports it. I've updated flutter tool.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -146,10 +149,203 @@ class WidgetInspectorService {
   ///
   /// Use this method only for testing to ensure that object references from one
   /// test case do not impact other test cases.
+  @visibleForTesting
```
I would make all these @protected, then just have a subclass in the tests, and have the subclass be the one you construct. You could even go further and require that the only way to extend this class is as a mixin, then you could hide the constructor on this version.

Hixie replied:
looks like you're actually a lot of the way there already actually.

jacob314 replied:
I've followed both those suggestions. The only way to extend is now a mixin and the tests are executed in a context where they can access protected members.

Hixie commented:
Sorry I didn't notice that you'd requested review. I think you requested review literally a few hours before I declared github mail bankruptcy and changed my settings to no longer watch the entire repo. :-(

jacob314 commented on packages/flutter/test/widgets/widget_inspector_test.dart:
```diff
@@ -134,536 +87,971 @@ void main() {
+      expect(await service.testExtension('getSelectedWidget', <String, String>{'objectGroup': 'my-group'}), contains('createdByLocalProject'));
+
+      await service.testExtension('setPubRootDirectories', <String, String>{'arg0': '/invalid/$pubRootTest'});
+      expect(await service.testExtension('getSelectedWidget', <String, St
```
The only delta from the previous commit other than indenting by two spaces as this is now a static method is the addition of this ext.flutter.inspector.show test.

jacob314 committed:
Fix bad import.

jacob314 commented:
PTAL

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -117,11 +122,15 @@ class _InspectorReferenceData {
-  static final WidgetInspectorService _instance = new WidgetInspectorService._();
+  static final WidgetInspectorService _instance = new _WidgetInspectorService();
+
+  bool _debugServiceExtensionsRegistered = false;
```
shouldn't this be static?

jacob314 replied:
Yes. Changed it to static and cleaned up how the WidgetInspectorService instance was set for
testing purposes so that the asserts don't fail.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -146,10 +155,259 @@ class WidgetInspectorService {
+  @protected
+  void registerServiceExtension({
+    @required String name,
+    @required FutureOr<Map<String, Object>> callback(Map<String, String> parameters),
```
might be worth making this a typedef to clarify what's going on

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -146,10 +155,259 @@ class WidgetInspectorService {
+  /// property named `result` in the JSON. In case of failure, the failure is
+  /// reported to the remote caller and is dumped to the logs.
+  @protected
+  void registerServiceExtension({
```
isn't this duplicating a lot of code from the binding layer?

jacob314 replied:
two methods are duplicated. The others support the specific API signature I've found convenient for the inspector of methods accepting two string valued arguments the first being an object id and the second an object group. cleaned up the most egregious duplicate code as discussed bellow.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -117,11 +122,15 @@ class _InspectorReferenceData {
   /// The current [WidgetInspectorService].
   static WidgetInspectorService get instance => _instance;
-  static final WidgetInspectorService _instance = new WidgetInspectorService._();
+  static final WidgetInspectorService _instance = new _WidgetInspectorService();
```
You could make the comment say that it's usable as a mixin for test purposes and as a singleton [instance] for production purposes, or something.

jacob314 replied:
Note that as soon as I add a private constructor to WidgetInspectorService I get a static analysis error  that I can't use it as a Mixin as it has a constructor. I made the factory constructor non-degenerate anyway and return a call to new _WidgetInspectorService() as maybe that is what you meant.

jacob314 commented:
I don't think making WidgetInspector a Binding makes sense and I think it is good addServiceExtension is a protected method on BindingBase
so I would propose making an addFlutterServiceExtension top level method
that both the Binding and WidgetInspector can forward to
to clean up that duplicated method
registerBoolServiceExtension would also be duplicated but the method is short and avoiding that duplication would cause more ugliness as it needs to forward to the overridable for tests registerServiceExtension call.

jacob314 committed:
Code review comment fixes.

jacob314 committed:
Avoid making registerFlutterServiceExtension public.
This has the benefit that if someone in the future creates a Binding
that dispatches service extensions via a completely different transfer
protocol (e.g. directly on a websocket)
the inspector service will also be transmitted over that protocol.

jacob314 committed:
Our extensions now get included in the count when service_extensions_test runs.

Hixie commented on packages/flutter/test/foundation/service_extensions_test.dart:
```diff
@@ -508,7 +508,7 @@ void main() {
     // If you add a service extension... TEST IT! :-)
     // ...then increment this number.
-    expect(binding.extensions.length, 17);
+    expect(binding.extensions.length, 32);
```
add a comment pointing to where the tests are for each one you added

jacob314 replied:
done

Hixie reviewed (commented):
LGTM

jacob314 committed:
Code review comment fix.
